matching-notification-migrations/src/main/java/org/phenotips/matchingnotification/internal/R74696PatientNetwork456DataMigration.java

```diff
@@ -0,0 +1,127 @@
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses/
+ */
+
+package org.phenotips.matchingnotification.internal;
+
+import org.xwiki.component.annotation.Component;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+
+import org.hibernate.HibernateException;
+import org.hibernate.Query;
+import org.hibernate.SQLQuery;
+import org.hibernate.Session;
+import org.hibernate.Transaction;
+import org.slf4j.Logger;
+
+import com.xpn.xwiki.XWikiException;
+import com.xpn.xwiki.store.hibernate.HibernateSessionFactory;
+import com.xpn.xwiki.store.migration.DataMigrationException;
+import com.xpn.xwiki.store.migration.XWikiDBVersion;
+import com.xpn.xwiki.store.migration.hibernate.AbstractHibernateDataMigration;
+
+/**
+ * Migration for PatientNetwork issue PN-456: local matches should be de-duplicated:
+ * given A-to-B and B-to-A matches, only A-to-B match should be left, where A < B alphanumericaly
+ * (to match how matching notification works after PN-401 is implemented).
+ *
+ * @version $Id$
+ * @since 1.3m1
+ */
+@Component
+@Named("R74696PatientNetwork456")
+@Singleton
+public class R74696PatientNetwork456DataMigration extends AbstractHibernateDataMigration
+{
+    // the "d.comments IS NULL" check is a safeguard, since in the production DB there should be no matches
+    // that have comments and have a notified copy of a match (and in fact there are none as of last check)
+    private static final String SQL_DELETE_NON_NOTIFIED_COPIES_OF_A_NOTIFIED_MATCH =
+        "delete from patient_matching d where d.comments IS NULL and d.notified = false and exists"
+        + " (select * from patient_matching d2 where"
+        + " d2.referencePatientId = d.referencePatientId and d2.matchedPatientId = d.matchedPatientId"
+        + " and d2.matchedServerId = d.matchedServerId and d2.referenceServerId = d.referenceServerId"
+        + " and d2.notified = true)";
+
+    private static final String HQL_DELETE_DUPLICATE_LOCAL_MATCHES =
+        "delete from CurrentPatientMatch d where d.matchedServerId = '' and d.referenceServerId = ''"
+        + " and d.referencePatientId > d.matchedPatientId and exists"
+        + " (from CurrentPatientMatch d2 where"
+        + " d2.referencePatientId = d.matchedPatientId and d2.matchedPatientId = d.referencePatientId"
+        + " and d2.matchedServerId = '' and d2.referenceServerId = '')";
+
+    private static final String NORMALIZE_ORDER_OF_PATIENTS_IN_A_MATCH =
+        "update CurrentPatientMatch set referencePatientId = matchedPatientId, matchedPatientId = referencePatientId"
```

Comment:
runs on all patient matches, has to either run on local matches only, or switch server IDs.